### ISSUE TITLE
The meaning of the SAMPLE_GROUPS field

### ISSUE THREAD

**Issue by learnuser1** [open-mmlab/OpenPCDet#1415]
In OpenPCDet dataset.yaml, What is the use of SAMPLE_GROUPS in DATA_AUGMENTOR, and what are the values in it based on? Do I need to make SAMPLE_GROUPS larger for a smaller number of objects if I'm training my data set

**Comment by SimonSongg** [comment]
I have the same confusion, any suggestion?

**Comment by AJAY31797** [comment]
@learnuser1 and @SimonSongg, I have the same confusion. Were you able to figure out what exactly they mean?

Thanks.

**Comment by SimonSongg** [comment]
> @learnuser1 and @SimonSongg, I have the same confusion. Were you able to figure out what exactly they mean?
> 
> Thanks.

Yes, in the data preparation process, all the target point cloud in the dataset are extracted and saved in "gt database". And in data augmentation step, these targets will be randomly copy paste into the scene. And these numbers in sampling group control how many targets in total you want in the scene. For example, if there's only one car in the scene, and you set the number for class CAR to be 5, then 4 more targets from other frames (in gt database) will be pasted into current scene. There are some security check to make sure the "pasted target" does not overlap with the exsiting target in the scene.

**Comment by AJAY31797** [comment]
Thanks @SimonSongg

**Comment by yinheyanxian** [comment]
> > 而且，我也有同样的困惑。你能弄清楚它们到底是什么意思吗？
> > 谢谢。
> 
> 是的，在数据准备过程中，数据集中的所有目标点云都会被提取并保存在“gt数据库”中。而在数据增强步骤中，这些目标将被随机复制粘贴到场景中。采样组中的这些数字控制着场景中想要的目标总数。例如，如果场景中只有一辆汽车，并且您将 CAR 类的数字设置为 5，则来自其他帧（在 gt 数据库中）的另外 4 个目标将粘贴到当前场景中。有一些安全检查，以确保“粘贴的目标”不会与场景中的现有目标重叠。

Hello, thank you for your answer. I noticed that there is also a parameter DISABLE_AUG_LIST: ['placeholder'] in DATA_AUGMENTOR, which seems to disable some data enhancement methods. I wonder what this parameter stands for.

    DATA_AUGMENTOR:
        DISABLE_AUG_LIST: ['placeholder']  # What does this parameter do?
        AUG_CONFIG_LIST:
            - NAME: gt_sampling
              USE_ROAD_PLANE: True
              DB_INFO_PATH:
                  - kitti_dbinfos_train.pkl
              PREPARE: {
                 filter_by_min_points: ['Car:5', 'Pedestrian:5', 'Cyclist:5'],
                 filter_by_difficulty: [-1],
              }

              SAMPLE_GROUPS: ['Car:15','Pedestrian:15', 'Cyclist:15']
              NUM_POINT_FEATURES: 4
              DATABASE_WITH_FAKELIDAR: False
              REMOVE_EXTRA_WIDTH: [0.0, 0.0, 0.0]
              LIMIT_WHOLE_SCENE: False

            - NAME: random_world_flip
              ALONG_AXIS_LIST: ['x']

            - NAME: random_world_rotation
              WORLD_ROT_ANGLE: [-0.78539816, 0.78539816]

            - NAME: random_world_scaling
              WORLD_SCALE_RANGE: [0.95, 1.05]

            - NAME: random_local_pyramid_aug
              DROP_PROB: 0.25
              SPARSIFY_PROB: 0.05
              SPARSIFY_MAX_NUM: 50
              SWAP_PROB: 0.1
              SWAP_MAX_NUM: 50

**Comment by yinheyanxian** [comment]
> > @learnuser1 and @SimonSongg, I have the same confusion. Were you able to figure out what exactly they mean?
> > Thanks.
> 
> Yes, in the data preparation process, all the target point cloud in the dataset are extracted and saved in "gt database". And in data augmentation step, these targets will be randomly copy paste into the scene. And these numbers in sampling group control how many targets in total you want in the scene. For example, if there's only one car in the scene, and you set the number for class CAR to be 5, then 4 more targets from other frames (in gt database) will be pasted into current scene. There are some security check to make sure the "pasted target" does not overlap with the exsiting target in the scene.

Hello, thank you for your answer. I noticed that there is also a parameter DISABLE_AUG_LIST: ['placeholder'] in DATA_AUGMENTOR, which seems to disable some data enhancement methods. I wonder what this parameter stands for.

    DATA_AUGMENTOR:
        DISABLE_AUG_LIST: ['placeholder']  # What does this parameter do?
        AUG_CONFIG_LIST:
            - NAME: gt_sampling
              USE_ROAD_PLANE: True
              DB_INFO_PATH:
                  - kitti_dbinfos_train.pkl
              PREPARE: {
                 filter_by_min_points: ['Car:5', 'Pedestrian:5', 'Cyclist:5'],
                 filter_by_difficulty: [-1],
              }

              SAMPLE_GROUPS: ['Car:15','Pedestrian:15', 'Cyclist:15']
              NUM_POINT_FEATURES: 4
              DATABASE_WITH_FAKELIDAR: False
              REMOVE_EXTRA_WIDTH: [0.0, 0.0, 0.0]
              LIMIT_WHOLE_SCENE: False

            - NAME: random_world_flip
              ALONG_AXIS_LIST: ['x']

            - NAME: random_world_rotation
              WORLD_ROT_ANGLE: [-0.78539816, 0.78539816]

            - NAME: random_world_scaling
              WORLD_SCALE_RANGE: [0.95, 1.05]

            - NAME: random_local_pyramid_aug
              DROP_PROB: 0.25
              SPARSIFY_PROB: 0.05
              SPARSIFY_MAX_NUM: 50
              SWAP_PROB: 0.1
              SWAP_MAX_NUM: 50

**Comment by Petros626** [comment]
> > @learnuser1 and @SimonSongg, I have the same confusion. Were you able to figure out what exactly they mean?
> > Thanks.
> 
> Yes, in the data preparation process, all the target point cloud in the dataset are extracted and saved in "gt database". And in data augmentation step, these targets will be randomly copy paste into the scene. And these numbers in sampling group control how many targets in total you want in the scene. For example, if there's only one car in the scene, and you set the number for class CAR to be 5, then 4 more targets from other frames (in gt database) will be pasted into current scene. There are some security check to make sure the "pasted target" does not overlap with the exsiting target in the scene.

SECOND paper:

3.2.2. Data Augmentation
Sample Ground Truths from the Database
The major problem we encountered during training was the existence of too few ground truths,
which significantly limited the convergence speed and final performance of the network. To solve this
problem, we introduced a data augmentation approach. First, we generated a database containing
the labels of all ground truths and their associated point cloud data (points inside the 3D bounding
boxes of the ground truths) from the training dataset. Then, during training, we randomly selected
several ground truths from this database and introduced them into the current training point cloud
via concatenation. Using this approach, we could greatly increase the number of ground truths per
point cloud and simulate objects existing in different environments. To avoid physically impossible
outcomes, we performed a collision test after sampling the ground truths and removed any sampled
objects that collided with other objects.

**Comment by Petros626** [comment]
> > > @learnuser1 and @SimonSongg, I have the same confusion. Were you able to figure out what exactly they mean?
> > > Thanks.
> > 
> > 
> > Yes, in the data preparation process, all the target point cloud in the dataset are extracted and saved in "gt database". And in data augmentation step, these targets will be randomly copy paste into the scene. And these numbers in sampling group control how many targets in total you want in the scene. For example, if there's only one car in the scene, and you set the number for class CAR to be 5, then 4 more targets from other frames (in gt database) will be pasted into current scene. There are some security check to make sure the "pasted target" does not overlap with the exsiting target in the scene.
> 
> Hello, thank you for your answer. I noticed that there is also a parameter DISABLE_AUG_LIST: ['placeholder'] in DATA_AUGMENTOR, which seems to disable some data enhancement methods. I wonder what this parameter stands for.
> 
> ```
> DATA_AUGMENTOR:
>     DISABLE_AUG_LIST: ['placeholder']  # What does this parameter do?
>     AUG_CONFIG_LIST:
>         - NAME: gt_sampling
>           USE_ROAD_PLANE: True
>           DB_INFO_PATH:
>               - kitti_dbinfos_train.pkl
>           PREPARE: {
>              filter_by_min_points: ['Car:5', 'Pedestrian:5', 'Cyclist:5'],
>              filter_by_difficulty: [-1],
>           }
> 
>           SAMPLE_GROUPS: ['Car:15','Pedestrian:15', 'Cyclist:15']
>           NUM_POINT_FEATURES: 4
>           DATABASE_WITH_FAKELIDAR: False
>           REMOVE_EXTRA_WIDTH: [0.0, 0.0, 0.0]
>           LIMIT_WHOLE_SCENE: False
> 
>         - NAME: random_world_flip
>           ALONG_AXIS_LIST: ['x']
> 
>         - NAME: random_world_rotation
>           WORLD_ROT_ANGLE: [-0.78539816, 0.78539816]
> 
>         - NAME: random_world_scaling
>           WORLD_SCALE_RANGE: [0.95, 1.05]
> 
>         - NAME: random_local_pyramid_aug
>           DROP_PROB: 0.25
>           SPARSIFY_PROB: 0.05
>           SPARSIFY_MAX_NUM: 50
>           SWAP_PROB: 0.1
>           SWAP_MAX_NUM: 50
> ```

A list of augmentation methods that are to be deactivated. In this case no data augmentation technique is disabled, it's just a placeholder

**Comment by Petros626** [comment]
> > @learnuser1 and @SimonSongg, I have the same confusion. Were you able to figure out what exactly they mean?
> > Thanks.
> 
> Yes, in the data preparation process, all the target point cloud in the dataset are extracted and saved in "gt database". And in data augmentation step, these targets will be randomly copy paste into the scene. And these numbers in sampling group control how many targets in total you want in the scene. For example, if there's only one car in the scene, and you set the number for class CAR to be 5, then 4 more targets from other frames (in gt database) will be pasted into current scene. There are some security check to make sure the "pasted target" does not overlap with the exsiting target in the scene.

@SimonSongg can you show me where this happens in the code: "these targets will be randomly copy paste into the scene."

**Comment by SimonSongg** [comment]
> > > @learnuser1 and @SimonSongg, I have the same confusion. Were you able to figure out what exactly they mean?
> > > Thanks.
> > 
> > 
> > Yes, in the data preparation process, all the target point cloud in the dataset are extracted and saved in "gt database". And in data augmentation step, these targets will be randomly copy paste into the scene. And these numbers in sampling group control how many targets in total you want in the scene. For example, if there's only one car in the scene, and you set the number for class CAR to be 5, then 4 more targets from other frames (in gt database) will be pasted into current scene. There are some security check to make sure the "pasted target" does not overlap with the exsiting target in the scene.
> 
> @SimonSongg can you show me where this happens in the code: "these targets will be randomly copy paste into the scene."

In OpenPCDet/pcdet/datasets/augmentor/database_sampler.py, `add_sampled_boxes_to_scene` function does this.

**Comment by Petros626** [comment]
Thx I found it just 1day before.

But what does image functions like `copy_paste_to_image_kitti()`, `collect_image_crops_kitti()` are purposed for when not using image augmentation in kitti_dataset.yaml, just for purpose? I try to isolate the code necessary for gt_sampling.

**Comment by Petros626** [comment]
> > > > @learnuser1 and @SimonSongg, I have the same confusion. Were you able to figure out what exactly they mean?
> > > > Thanks.
> > > 
> > > 
> > > Yes, in the data preparation process, all the target point cloud in the dataset are extracted and saved in "gt database". And in data augmentation step, these targets will be randomly copy paste into the scene. And these numbers in sampling group control how many targets in total you want in the scene. For example, if there's only one car in the scene, and you set the number for class CAR to be 5, then 4 more targets from other frames (in gt database) will be pasted into current scene. There are some security check to make sure the "pasted target" does not overlap with the exsiting target in the scene.
> > 
> > 
> > @SimonSongg can you show me where this happens in the code: "these targets will be randomly copy paste into the scene."
> 
> In OpenPCDet/pcdet/datasets/augmentor/database_sampler.py, `add_sampled_boxes_to_scene` function does this.

@SimonSongg where is the collision avoidance implemented, when inserting new objects and is it also used for techqniues like local_rotation or translation?

**Comment by SimonSongg** [comment]
> > > > > @learnuser1 and @SimonSongg, I have the same confusion. Were you able to figure out what exactly they mean?
> > > > > Thanks.
> > > > 
> > > > 
> > > > Yes, in the data preparation process, all the target point cloud in the dataset are extracted and saved in "gt database". And in data augmentation step, these targets will be randomly copy paste into the scene. And these numbers in sampling group control how many targets in total you want in the scene. For example, if there's only one car in the scene, and you set the number for class CAR to be 5, then 4 more targets from other frames (in gt database) will be pasted into current scene. There are some security check to make sure the "pasted target" does not overlap with the exsiting target in the scene.
> > > 
> > > 
> > > @SimonSongg can you show me where this happens in the code: "these targets will be randomly copy paste into the scene."
> > 
> > 
> > In OpenPCDet/pcdet/datasets/augmentor/database_sampler.py, `add_sampled_boxes_to_scene` function does this.
> 
> @SimonSongg where is the collision avoidance implemented, when inserting new objects and is it also used for techqniues like local_rotation or translation?

1. In `__call__`, it will calculate the 3d IoU of sampled bbox and existing bbox to check if collision exists. 2. It seems not.

**Comment by Petros626** [comment]
@SimonSongg I agree with the first one. The objects e.g. in KITTI can't overlap by local_rotation, because all object are rotated over the yaw(z)-axis, correct? There is no independent rotation of objects, that a rotation collision could appear.

**Comment by SimonSongg** [comment]
> @SimonSongg I agree with the first one. Are the objects e.g. in KITTI so far away that there is no overlap in rotation? For example only a high rotation like 45° could intersect with another object.

I dont get you point.

**Comment by Petros626** [comment]
> > @SimonSongg I agree with the first one. Are the objects e.g. in KITTI so far away that there is no overlap in rotation? For example only a high rotation like 45° could intersect with another object.
> 
> I dont get you point.

Updated my previous question above.

**Comment by SimonSongg** [comment]
> @SimonSongg I agree with the first one. The objects e.g. in KITTI can't overlap by local_rotation, because all object are rotated over the yaw(z)-axis, correct? There is no independent rotation of objects, that a rotation collision could appear.

I don't see local rotation of objects. Still confused about your question.

**Comment by Petros626** [comment]
For example when I use the augmentation technique `random_local_rotation` for my KITTI Dataset,there is no check function like for gt_sampling. For gt_sampling `__call__` with handle this, but when I'm using `random_local_rotation` how it's handled there?

**Comment by Petros626** [comment]
> For example when I use the augmentation technique `random_local_rotation` for my KITTI Dataset,there is no check function like for gt_sampling. For gt_sampling `__call__` with handle this (collision avoidance), but when I'm using `random_local_rotation` how it's handled.

**Comment by SimonSongg** [comment]
> For example when I use the augmentation technique `random_local_rotation` for my KITTI Dataset,there is no check function like for gt_sampling. For gt_sampling `__call__` with handle this, but when I'm using `random_local_rotation` how it's handled there?

It is not used by default. So you may need to handle collision by yourself.

**Comment by Petros626** [comment]
But when all objects and boxes in the point cloud scene are locally rotated by e.g. 9°, then there is no collision, because all objects are rotated and not some overall objects.

**Comment by SimonSongg** [comment]
> But when all objects and boxes in the point cloud scene are locally rotated by e.g. 9°, then there is no collision, because all objects are rotated and not some overall objects.

Your statement is wrong.
![2f6d0e561f08bd028d751e172e848fd](https://github.com/user-attachments/assets/b12df4fe-a255-4a28-8b24-231e12095674)

**Comment by Petros626** [comment]
brown = bb before rotation and blue = bb after rotation. Like I mentioned the angle shouldn't be high (only 10°) when rotation locally objects that there is no intersection between objects.

**Comment by SimonSongg** [comment]
> brown = objects within points? and blue = original position?

Bounding box before rotation and after rotation.

**Comment by Petros626** [comment]
Okay so it's depending on the rotation angle. A small value could working without collision.

**Comment by SimonSongg** [comment]
> Okay so it's depending on the rotation angle. A small value could working without collision.

As you like... If two objects are very close to each other, they will collide even if 1 degree rotation though.

**Comment by Petros626** [comment]
> > Okay so it's depending on the rotation angle. A small value could working without collision.
> 
> As you like... If two objects are very close to each other, they will collide even if 1 degree rotation though.

Why the authors didn't consider that during the development? Did you used `local_rotation` when you was working with OpenPCDet? 

@SimonSongg could you guide me here, I've implemented a strategy. OBB should be the correct one for 3D LiDAR BB.
